File: p523/p523.go
// Brutal.
// Note that k can be zero.

package p523

func checkSubarraySum(nums []int, k int) bool {
	n := len(nums)

	for i := 0; i < n-1; i++ {
		sum := nums[i]

		for j := i + 1; j < n; j++ {
			sum += nums[j]

			if (k == 0 && sum == 0) || (k != 0 && sum%k == 0) {
				return true
			}
		}
	}
	return false
}

File: p523/p523.cpp
class Solution {
public:
    bool checkSubarraySum(vector<int>& nums, int k) {
       	unordered_set<int> setOfSums;
       	int sum = 0, pre = 0;

       	for (int i = 0; i < nums.size(); ++i)
       	{
       		sum += nums[i];
       		if (k != 0)
       			sum %= k;

       		if (setOfSums.count(sum)) {
       			return true;
       		}

       		setOfSums.insert(pre);
       		pre = sum;
       	}

       	return false;
    }
};